### ISSUE TITLE
New aggregate functions FIRST and LAST

### ISSUE THREAD

**Issue by blueedgenick** [confluentinc/ksql#1128]
Feature request to add these 2 new aggregate functions, which should return the earliest or latest (based on rowtime) values for a specified column from the aggregate window. In the case of a non-windowed query the implicit window is from the dawn of time (1/1/1970) onwards.
These simple functions can unlock a lot of valuable use-cases, especially when used with `SESSION` windows - e.g. what was the last product details page visited before abandoning a shopping cart ? There are plenty of similar examples for basket analysis, fraud detection, error monitoring, and so on besides website analytics.

Notice that unlike many aggregate functions these should work on any datatype, not just numerics.

One possible complexity for implementation is that the current KUDAF skeleton doesn't expose the source record so the timestamp won't be visible. One possible approach might be to include the name of the timestamp field to use for evaluation as a second argument e.g. `FIRST(metric_col, ts_col)`, which would have the added flexibility of allowing evaluation based on another long field in the record (although that could also be confusing, given that actual windowing semantics have to be based on ROWTIME).

**Comment by dguy** [comment]
@blueedgenick if it is based on `ROWTIME` why do we need the additional `ts_col`? The `ROWTIME` is set to either the record timestamp or to the value of the field defined as the timestamp field, so would that not be the correct field to use for this function?

**Comment by blueedgenick** [comment]
@dguy you are correct that, semantically, it only makes sense to put ROWTIME there - i was just brainstorming a way to more easily make the rowtime available to the actual KUDAF implementation. Talking it through with @hjafarpour he made a smarter suggestion along the lines of hiding that detail from the user - so she only has to type `FIRST(my_first_col)` - and then leverage the newly-added parent node reference in the AST to do an "internal query rewrite" to add the ROWTIME as a second arg to the internal KUDAF implementation. (it's to do with the way we do the code-gen for the functions). At that point I stopped thinking i would contribute this one myself on a flight somewhere and instead leave it to you folks who know what you're doing ;)

**Comment by lijia1** [comment]
I vote for this feature too and prefer @vjache 's ARGMAX() approach. Ranking using a user-supplied column is better because potentially the values in this column are unique and thus tie-breaking is not necessary. The same cannot be said if ROWTIME is used.

**Comment by kalyanmysore** [comment]
I think having a LAST feature would be very useful for creating streams that generate samples of another. Eg: If I need to send updates to a mobile device once every 60 sec, using the last known state of some STRUCT data that I'm building up.
Without this, I'd have to manage that state and the timer outside of kafka. If you implement the feature, I'd just have to have a connector (and maybe some small amount of code).

**Comment by blueedgenick** [comment]
Well, if your problem statement is "continuously update a state object from
incoming records and send the latest value of that state downstream every n
seconds" you could do that today with the Kafka Streams library.

Doing it in KSQL with a LAST (or any other) aggregation wouldn't actually
do what you need because the aggregate values are emitted downstream every
time they change in response to an input record, whereas you only want them
every minute. I suggest you take a look at #1030 and upvote there if it
would address your usecase

**Comment by big-andy-coates** [comment]
Closing as a duplicate of #3985

**Comment by blueedgenick** [comment]
i'm going to re-open this as it's actually something different than #3985 (and even if it were the same, i generally prefer retaining the older issue and closing the newer 'dupe').
